python/metatensor-learn/metatensor/learn/nn/module_map.py

```diff
@@ -0,0 +1,470 @@
+        """
+        Takes a tensor map and applies the modules on each key it.
+
+        :param tensor:
+            input tensor map
+        """
+        out_blocks: List[TensorBlock] = []
+        for key, block in tensor.items():
+            out_block = self.forward_block(key, block)
+
+            for parameter, gradient in block.gradients():
+                if len(gradient.gradients_list()) != 0:
+                    raise NotImplementedError(
+                        "gradients of gradients are not supported"
+                    )
+                out_block.add_gradient(
+                    parameter=parameter,
+                    gradient=self.forward_block(key, gradient),
+                )
+            out_blocks.append(out_block)
+
+        return TensorMap(tensor.keys, out_blocks)
+
+    def forward_block(self, key: LabelsEntry, block: TensorBlock) -> TensorBlock:
+        position = self._in_keys.position(key)
+        if position is None:
+            raise KeyError(f"key {key} not found in modules.")
+        module_idx: int = position
+        module: ModuleMapInterface = self._modules_list[module_idx]
+        out_values = module.forward(block.values)
+        if self._out_tensor is None:
+            properties = Labels.range("_", out_values.shape[-1])
+        else:
+            properties = self._out_tensor.block(key).properties
+        return TensorBlock(
+            values=out_values,
+            properties=properties,
+            components=block.components,
+            samples=block.samples,
+        )
+
+    @property
+    def modules(self):
+        """
+        :return modules:
+            torch.nn.ModulesList that was given by initialization. It is not a copy, so
+            its utility functionalities can be used to e.g. move the module map to GPU
+        """
+        # type annotation in function signature had to be removed because of TorchScript
+        return self._modules_list
+
+    @torch.jit.export
+    def get_module(self, key: LabelsEntry):
+        """
+        :param key:
+            key of module which should be returned
+
+        :return module:
+            returns he torch.nn.Module corresponding to the :param key:
+        """
+        # type annotation in function signature had to be removed because of TorchScript
+        position = self._in_keys.position(key)
+        if position is None:
+            raise KeyError(f"key {key} not found in modules.")
+        module_idx: int = position
+        module: ModuleMapInterface = self._modules_list[module_idx]
+        return module
+
+    @property
+    def in_keys(self) -> Labels:
+        """ """
+        return self._in_keys
+
+    @property
+    def out_tensor(self) -> Optional[TensorMap]:
+        """
+        The tensor map that is used to determine properties labels of the output of
+        forward function.
+        """
+        return self._out_tensor
+
+
+class Linear(ModuleMap):
```

Comment:
Okay, here I'm not sure about the general idea of how it works. It seems like if we want to use a `Linear` module, we should not only create an input `TensorMap`, but also an output `TensorMap` to provide desired dimensions size. IMO, creating tensormaps is really something painful, particularly if we want to stack a few `Linear` modules together.

So it's probably better to have the closest possible to the `torch.nn.Linear` implementation with explicit dimensions size, because this is something people would expect

Reply:
It is quite close to torch. Torch requires also to give input and output dimension, Because we can have different input and output dimensions for each tensor block, we need at least these information for each block. This I can adapt as it was suggested for the output labels https://github.com/lab-cosmo/metatensor/pull/427/files#r1452473183 by merging the two constructors (`__init__` and `from_module`)

```python
class Linear
    def __init__(
        self,
        in_properties:  Union[int, Dict[LabelsEntry, Labels]],
        out_properties: Union[int, Dict[LabelsEntry, Labels]],
        bias: Union[bool, Dict[LabelsEntry, bool]] = True,
        device: torch.device = None,
        dtype: torch.dtype = None,
        many_to_one: bool = True,
    ):

```

What do you think?

Reply:
I agree that in general the different size of the blocks may be an issue, but I feel like in my applications it have never happened. 

Let say, we never want to use the Linear layer on top of the raw spherical expansion coefficients, because we then mix channels and break everything. But usually in both radial and power spectrum, the number of features in each block is fixed, and thus providing only input and output sizes once was working perfectly.

Reply:
> I agree that in general the different size of the blocks may be an issue, but I feel like in my applications it have never happened.

In equivariant learning it happens. But both will ways will be supported. If we have an example for the easy case (as first one) and one for the complex case then it should be fine for the user to understand quickly the more complicated function signature.

Reply:
Okay, then maybe it's better to just provide a list or tuple of input and output dimensions? 
My point is, creating a tensormap by hands is usually very inconvenient, and it requires the list of dimensions anyway. What do you think of

```py
class Linear(ModuleMap):
    def __init__(self, in_features: Union[List, Tuple], out_features: Union[List, Tuple], bias: Bool = True, device = None, dtype = None):
        pass
```

And then just initialising the set of Linear layers for each block separately?

Reply:
I'm not sure what you are suggesting here. I agree that building TensorMap by hand is a bit clunky, but it is not required by the latest API suggestion? For `in_properties:  Union[int, Dict[LabelsEntry, Labels]]`, you can use it as

```py
linear = Linear(in_properties=3, out_properties=6)
```

or if you care

```py
linear = Linear(in_properties=3, out_properties=Labels("nn_features", [1, 2, 5, 4]))
```

or if you want different things for each block

```py
linear = Linear(in_properties=[3, 4, 5], out_properties=[Labels(...), Labels(...), Labels(...)])
```

I don't understand what would be inside the `List` or `Tuple` in your suggestion